waves/src/ConfirmSwapDrawer.tsx

```diff
@@ -29,24 +29,78 @@ interface ConfirmSwapDrawerProps {
     onCancel: () => void;
     onSwapped: (txId: string) => void;
     transaction: string;
+    btcBalance: number;
+    usdtBalance: number;
 }
 
-export default function ConfirmSwapDrawer({ isOpen, onCancel, onSwapped, transaction }: ConfirmSwapDrawerProps) {
-    // eslint-disable-next-line @typescript-eslint/no-unused-vars
-    let { data, error } = useSWR("decompose-transaction", () => decomposeTransaction(transaction));
+export default function ConfirmSwapDrawer(
+    { isOpen, onCancel, onSwapped, transaction, btcBalance, usdtBalance }: ConfirmSwapDrawerProps,
+) {
+    const [alphaAsset, setAlphaAsset] = useState<AssetType>(AssetType.BTC);
+    const [alphaAmount, setAlphaAmount] = useState(0);
+    const [alphaBalanceBefore, setAlphaBalanceBefore] = useState(0);
+    const [alphaBalanceAfter, setAlphaBalanceAfter] = useState(0);
+    const [betaAsset, setBetaAsset] = useState<AssetType>(AssetType.USDT);
+    const [betaAmount, setBetaAmount] = useState(0);
+    const [betaBalanceBefore, setBetaBalanceBefore] = useState(0);
+    const [betaBalanceAfter, setBetaBalanceAfter] = useState(0);
+
+    function tickerToAssetType(ticker: string) {
+        switch (ticker) {
+            case LBTC_TICKER:
+                return AssetType.BTC;
+            case LUSDT_TICKER:
+                return AssetType.USDT;
+            default:
+                throw new Error("Unexpected ticker: " + ticker);
+        }
+    }
+
+    useAsyncEffect(async () => {
```

Comment:
Instead of using `useAsyncEffect` and several `useStates` how about:

- 1 `useAsync` with `promiseFn` (that will make it execute immediately instead of having to trigger it with `run`)
- You can use the `watch` property to re-execute if `transaction` changes
- You can use the `isPending` property to check if the computation is done and display a loading spinner where you would normally display the balances while it is computing
- You can also chain all the functions like `tickerToAssetType` onto the `decomposeTransaction` like this:

```ts
useAsync({
	promiseFn: async () => {
		let transactionBalance = await decomposeTransaction(transaction);

		// do all your things here

		return {
			alphaAsset: ...,
			alphaAmount: ...,
			...
		}
	}
})
```